### ISSUE TITLE
Naming of sentinel values not present in JS

### ISSUE THREAD

**Issue by kainino0x** [webgpu-native/webgpu-headers#234]
In the C API we have a few sentinel enum values that don't have equivalents in JS:

- `WGPUVertexStepMode_VertexBufferNotUsed` (JS has `sequence<GPUVertexBufferLayout?>`) (see #242)
- `WGPUTextureFormat_Undefined` in `WGPUFragmentState` (#160) (JS has `sequence<GPUColorTargetState?>`)
- The binding types: BufferBindingType, SamplerBindingType, TextureSampleType, StorageTextureAccess `_Undefined` (JS has the parent dictionary optional) (see #242)
- `WGPUCompositeAlphaMode_Auto`

We also have `Undefined` variants for a bunch of enums which are used as optional-without-default in JS.

And finally, constants for sentinel values of numeric fields:

```c
#define WGPU_ARRAY_LAYER_COUNT_UNDEFINED (0xffffffffUL)
#define WGPU_COPY_STRIDE_UNDEFINED (0xffffffffUL)
#define WGPU_LIMIT_U32_UNDEFINED (0xffffffffUL)
#define WGPU_LIMIT_U64_UNDEFINED (0xffffffffffffffffULL)
#define WGPU_MIP_LEVEL_COUNT_UNDEFINED (0xffffffffUL)
#define WGPU_QUERY_SET_INDEX_UNDEFINED (0xffffffffUL)
#define WGPU_WHOLE_MAP_SIZE SIZE_MAX
#define WGPU_WHOLE_SIZE (0xffffffffffffffffULL)
```

Should we call all of them "undefined"?
Or should we be trying to use more meaningful names, like `VertexBufferNotUsed` and `WHOLE_SIZE`?
Or should we have both?

Examples of meaningful names:
- `WGPUTextureFormat_ColorTargetNotUsed`
- `WGPUBufferBindingType_NotUsed`
- `WGPU_ARRAY_LAYER_COUNT_DEFAULT` or e.g. `TO_END` or `ALL_REMAINING`
- `WGPU_QUERY_SET_INDEX_NOT_WRITTEN`

**Comment by kainino0x** [comment]
Found another: WGPUCompositeAlphaMode_Auto

**Comment by kainino0x** [comment]
I promised I'd make a proposal. Here's... sort of that. changes from current header marked with \*.

Firstly _just in case_ WGPUTextureFormat_Undefined ever becomes ambiguous with with an actual optional texture format in the API, we should probably go ahead and change it to `WGPUTextureFormat_ColorTargetNotUsed` > 0. Bit odd because we'll have this sentinel value just for one user of WGPUTextureFormat, but whatever. (https://github.com/webgpu-native/webgpu-headers/issues/242#issuecomment-1776324738)

Here are names I'd use if naming everything from scratch:

`WGPUVertexStepMode_VertexBufferNotUsed`
\* `WGPUTextureFormat_ColorTargetNotUsed`
\* `{BufferBindingType,SamplerBindingType,TextureSampleType,StorageTextureAccess}_BindingNotUsed` = 0
`WGPUCompositeAlphaMode_Auto` = 0
\* `WGPU_{ARRAY_LAYER_COUNT,MIP_LEVEL_COUNT}_MAX`
\* ~`WGPU_COPY_STRIDE_UNDEFINED`~ `WGPU_{BYTES_PER_ROW,ROW_PER_IMAGE}_UNDEFINED`. Easier to figure out which constant goes with these fields, but this should be solved with documentation anyway
`WGPU_LIMIT_{U32,U64}_UNDEFINED`. I looked for another name but DEFAULT is a little odd because it makes it sound like a constant, not a sentinel.
\* `WGPU_QUERY_SET_INDEX_NOT_WRITTEN`
\* `WGPU_BUFFER_MAP_RANGE_MAX`
\* `WGPU_BUFFER_RANGE_MAX`

However this is annoying in practice, so sticking to just things that I think really significantly enhance readability:

`WGPUVertexStepMode_VertexBufferNotUsed`
\* `WGPUTextureFormat_ColorTargetNotUsed`
\* `{BufferBindingType,SamplerBindingType,TextureSampleType,StorageTextureAccess}_BindingNotUsed` = 0. Almost no one is actually typing these, so renaming them is easy, but also not that important.
`WGPUCompositeAlphaMode_Auto` = 0
`WGPU_{ARRAY_LAYER_COUNT,MIP_LEVEL_COUNT}_UNDEFINED`. OK because matches JS API
`WGPU_COPY_STRIDE_UNDEFINED`
`WGPU_LIMIT_{U32,U64}_UNDEFINED`. Though should be low usage so easier to change.
`WGPU_QUERY_SET_INDEX_UNDEFINED`. OK because unambiguous, and matches JS API, but would be easy to change
`WGPU_WHOLE_MAP_SIZE`
`WGPU_WHOLE_SIZE`

**Comment by Kangz** [comment]
What do you think of making `Undefined` and `NotUsed` alias the same value? So it's just for semantics that the names are different?

**Comment by kainino0x** [comment]
I considered that for the binding types, but it doesn't seem any better to me than having just one or the other.

I don't want to do it for VertexBufferNotUsed and ColorTargetNotUsed, because of #242, though technically both could have their values reused for other sentinel values (other than Undefined) if we ever have them.

**Comment by kainino0x** [comment]
Actually I think I also want to number BindingNotUsed > 0 for consistency (posted on #242).

**Comment by kainino0x** [comment]
Nov 16 meeting agreement: do that

**Comment by yig** [comment]
For a user of the C API, it really helps to have default values map to 0. This allows one to zero-initialize a struct rather than set fields to non-zero values explicitly.

**Comment by Kangz** [comment]
That has always been a consideration, but there are a lot of cases where we need to make the difference between 0 and a value that's not set in Javascript, which is doable there. In a C API we cannot do it or it would require a nullable pointer to a value which would be even worse. If you use C++ these default values can be set for you by the C++ headers.

**Comment by austinEng** [comment]
for C, we also have the INIT structs, so you can do something like `WGPUTextureDescriptor desc = WGPU_TEXTURE_DESCRIPTOR_INIT;` to get the same defaults. it is mostly 0-initializing, but there are some cases we can't do that.

**Comment by yig** [comment]
That's nice. For most of WebGPU, I can do Sokol-style initialization <https://github.com/floooh/sokol/blob/master/sokol_gfx.h#L547> using designated initializers, as in:

```c++
WGPURenderPassEncoder render_pass = wgpuCommandEncoderBeginRenderPass( encoder, to_ptr(WGPURenderPassDescriptor{
    .colorAttachmentCount = 1,
    .colorAttachments = to_ptr(WGPURenderPassColorAttachment{{
        .view = current_texture_view,
        .depthSlice = WGPU_DEPTH_SLICE_UNDEFINED,
        .loadOp = WGPULoadOp_Clear,
        .storeOp = WGPUStoreOp_Store,
        // Choose the background color.
        .clearValue = WGPUColor{ .5, .5, .1, 1.0 }
        }})
    }) );
```

It looks almost as good as JavaScript. Non-zero defaults get in the way.

**Comment by kainino0x** [comment]
About `WGPUCompositeAlphaMode_Auto` in particular, I don't think it was particularly intentional to [renumber it](https://dawn-review.googlesource.com/c/dawn/+/192867) from 0 to 1 in Dawn (@austinEng?). My proposal earlier in this thread had it as 0 (because it behaves just like any other defaulting which we normally would call "Undefined").

There are definitely places where it's impractical to have the default be 0, but since it _is_ annoying, we should try to do it where we can (and this is part of why we renumbered enums so that zero-init wouldn't cause defaulting where there shouldn't be any defaulting).

**Comment by kainino0x** [comment]
(that's a reply to the longer [chat thread](https://matrix.to/#/!MFogdGJfnZLrDmgkBN:matrix.org/$kvSQZsvHqggfTwAGt9PMIkA5VOBxyfTaWEMlBONJW9Y?via=matrix.org&via=mozilla.org&via=kit.edu) about this)

**Comment by austinEng** [comment]
I think that was unintentional, yes. It can stay 0 as the default.

**Comment by kainino0x** [comment]
👍 https://dawn-review.googlesource.com/c/dawn/+/204174

**Comment by yig** [comment]
Great! I think `WGPUPresentMode_Fifo` needs the same treatment (value 0).

**Comment by kainino0x** [comment]
That one will be solved in a slightly different way, see #289
